mongodbatlas/fw_data_source_mongodbatlas_atlas_user.go

```diff
@@ -0,0 +1,254 @@
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	if atlasUserConfig.UserID.IsNull() && atlasUserConfig.Username.IsNull() {
+		resp.Diagnostics.AddError(errorMissingAttributesSummary, "either user_id or username must be configured")
+		return
+	}
+
+	var (
+		err  error
+		user *admin.CloudAppUser
+	)
+	if !atlasUserConfig.UserID.IsNull() {
+		userID := atlasUserConfig.UserID.ValueString()
+		user, _, err = connV2.MongoDBCloudUsersApi.GetUser(ctx, userID).Execute()
+		if err != nil {
+			resp.Diagnostics.AddError("error when getting User from Atlas", fmt.Sprintf(errorUserRead, userID, err.Error()))
+			return
+		}
+	} else {
+		username := atlasUserConfig.Username.ValueString()
+		user, _, err = connV2.MongoDBCloudUsersApi.GetUserByUsername(ctx, username).Execute()
+		if err != nil {
+			resp.Diagnostics.AddError("error when getting User from Atlas", fmt.Sprintf(errorUserRead, username, err.Error()))
+			return
+		}
+	}
+
+	userResultState := newTFAtlasUserDSModel(user)
+	resp.Diagnostics.Append(resp.State.Set(ctx, &userResultState)...)
+}
+
+func newTFAtlasUserDSModel(user *admin.CloudAppUser) tfAtlasUserDSModel {
+	return tfAtlasUserDSModel{
+		ID:           types.StringPointerValue(user.Id),
+		UserID:       types.StringPointerValue(user.Id),
+		Username:     types.StringValue(user.Username),
+		Country:      types.StringValue(user.Country),
+		CreatedAt:    types.StringPointerValue(TimePtrToStringPtr(user.CreatedAt)),
+		EmailAddress: types.StringValue(user.EmailAddress),
+		FirstName:    types.StringValue(user.FirstName),
+		LastAuth:     types.StringPointerValue(TimePtrToStringPtr(user.LastAuth)),
+		LastName:     types.StringValue(user.LastName),
+		MobileNumber: types.StringValue(user.MobileNumber),
+		TeamIDs:      user.TeamIds,
+		Links:        newTFLinksList(user.Links),
+		Roles:        newTFRolesList(user.Roles),
+	}
+}
+
+func newTFLinksList(links []admin.Link) []tfLinkModel {
+	if links == nil {
+		return nil
+	}
+	resLinks := make([]tfLinkModel, len(links))
+	for i, value := range links {
+		resLink := tfLinkModel{
+			Href: types.StringPointerValue(value.Href),
+			Rel:  types.StringPointerValue(value.Rel),
+		}
+		resLinks[i] = resLink
+	}
+	return resLinks
+}
+
+func newTFRolesList(roles []admin.CloudAccessRoleAssignment) []tfAtlasUserRoleModel {
+	if roles == nil {
+		return nil
+	}
+	resRoles := make([]tfAtlasUserRoleModel, len(roles))
+	for i, value := range roles {
+		resRole := tfAtlasUserRoleModel{
+			GroupID:  types.StringPointerValue(value.GroupId),
+			OrgID:    types.StringPointerValue(value.OrgId),
+			RoleName: types.StringPointerValue(value.RoleName),
+		}
+		resRoles[i] = resRole
+	}
+	return resRoles
+}
+
+// utility functions that can be places in sdk
+func TimePtrToStringPtr(t *time.Time) *string {
+	if t == nil {
+		return nil
+	}
+	res := t.String()
+	return &res
+}
+
+func int64PtrToIntPtr(i64 *int64) *int {
+	if i64 == nil {
+		return nil
+	}
+
+	i := int(*i64)
+	return &i
+}
```

Comment:
let's move this to the provider.go for now so that it is easier to find while we wait for the SDK to support ir

Reply:
makes sense, would however favor defining a util package for these simple conversions.

Reply:
Something to consider https://www.adam-bien.com/roller/abien/entry/util_packages_are_evil 😈